### PR TITLE
Scan build warnings 2

### PR DESCRIPTION
**Pull request by ryanhsu19**
Third-party files have been removed from the list. See #2379 for details.

### CODE REVIEW

File: samples/file-encryptor/enclave/keys.cpp [hunk 1]
```diff
@@ -290,9 +290,17 @@ int ecall_dispatcher::prepare_encryption_header(
     {
         TRACE_ENCLAVE("password_key");
         for (unsigned int i = 0; i < ENCRYPTION_KEY_SIZE_IN_BYTES; i++)
+        {
+            if (password_key[i] > UINT32_MAX)
```
Comment (jhand2) [inline samples/file-encryptor/enclave/keys.cpp]: Discussed previously, but it is not possible for this condition to be true. What warning is this addressing?

File: samples/file-encryptor/enclave/keys.cpp [hunk 2]
```diff
@@ -290,9 +290,17 @@ int ecall_dispatcher::prepare_encryption_header(
...
+            {
+                TRACE_ENCLAVE(
+                    "password key is too large to fit into an unsigned int", 1);
+                goto exit;
+            }
             TRACE_ENCLAVE(
                 "password_key[%d] =0x%02x", i, (unsigned int)(password_key[i]));
-        goto exit;
+            goto exit;
```
Comment (jhand2) [inline samples/file-encryptor/enclave/keys.cpp]: I think this goto should be outsize the loop

File: tests/crypto/asn1_tests.c [hunk 3]
```diff
@@ -264,8 +264,6 @@ static oe_result_t _parse(oe_asn1_t* asn1)
                 if (result != OE_OK)
                     goto done;
 
-                result = _parse(&sequence);
```
Comment (jhand2) [inline tests/crypto/asn1_tests.c]: I think this line is supposed to be kept, it is trying to recursively parse sequence.

The missing thing here is that the response isn't used. So the whole case should be:
```
case OE_ASN1_TAG_SEQUENCE:
{
    oe_asn1_t sequence;

    result = oe_asn1_get_sequence(asn1, &sequence);
    if (result != OE_OK)
        goto done;

    result = _parse(&sequence);
    if (result != OE_OK)
        goto done;

    break;
}
```

File: tests/libunwind/enc/enc.c [hunk 4]
```diff
@@ -65,10 +65,6 @@ int test(char test_name[201], uint32_t pid)
     int rval = 1;
     g_pid = pid;
     printf("RUNNING: %s\n", __TEST__);
-    if (!(__environ = (char**)calloc(1, sizeof(char**))))
-    {
-        rval = 1;
-    }
```
Comment (jhand2) [inline tests/libunwind/enc/enc.c]: This check should be kept. Just change `rval = 1;` to `return 1;`

**Comment by ryanhsu19** [comment]
Thank you! I will keep working on other bugs.

**Comment by radhikaj** [comment]
bors try

File: enclave/core/sgx/td.c [hunk 5]
```diff
@@ -91,19 +91,23 @@ void td_push_callsite(td_t* td, Callsite* callsite)
...
 {
     if (!td || !td->callsites)
-        oe_abort();
-
-    if (td->depth == 1)
     {
-        // The outermost ecall is about to return.
-        // Clear the thread-local storage.
-        td_clear(td);
+        oe_abort();
```
[the remark this answers is not in the document]
Comment (jhand2) [reply enclave/core/sgx/td.c]: scan-build doesn't know that `oe_abort` stops execution. So it thinks td is potentially null and that we just ignore the result of the check.

**Comment by jhand2** [comment]
Hey @ryanhsu19 can you clean up the commit history here? We shouldn't have the merge commit or the extra "check for unsigned int added by Ryan Hsu 12/6/2019" commit. The other two commits should just be squashed into one.

Let me know if you need help with how to do this.

File: enclave/core/sgx/td.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (mingweishih) [reply enclave/core/sgx/td.c]: I see. This is unfortunate as I like the original pattern better :)

File: enclave/core/sgx/report.c [hunk 6]
```diff
@@ -210,6 +210,9 @@ oe_result_t oe_get_remote_report(
 
     sgx_quote = (sgx_quote_t*)report_buffer;
 
+    if (sgx_quote == NULL)
+        OE_RAISE(OE_UNEXPECTED);
```
Comment (jhand2) [inline enclave/core/sgx/report.c]: We should raise `OE_INVALID_PARAMETER` instead of unexpected

File: enclave/sgx/revocationinfo.c [hunk 7]
```diff
@@ -81,20 +82,30 @@ oe_result_t oe_get_revocation_info(oe_get_revocation_info_args_t* args)
...
         {
-            if (!(in.tcb_info = realloc(in.tcb_info, out.tcb_info_size)))
+            if (!(temp = realloc(in.tcb_info, out.tcb_info_size)))
+            {
                 OE_RAISE(OE_OUT_OF_MEMORY);
+            }
+            else
+            {
+                in.tcb_info = temp;
+            }
```
Comment (jhand2) [inline enclave/sgx/revocationinfo.c]: What is the reason for doing this? There is no memory safety issue in the original implementation that i can see.

File: host/sgx/elf.c [hunk 8]
```diff
@@ -1674,7 +1677,10 @@ int elf64_add_section(
         /* Update number of sections */
         elf64_ehdr_t* ehdr = _get_header(elf);
         if (oe_safe_add_u16(ehdr->e_shnum, 1, &ehdr->e_shnum) != OE_OK)
+        {
+            mem_free(&mem);
```
Comment (jhand2) [inline host/sgx/elf.c]: This should just be done once after the `done` label

File: tests/libc/enc/enc.c [hunk 9]
```diff
@@ -111,7 +111,7 @@ int run_test(const char* name, int (*main)(int argc, const char* argv[]))
 
     /* Allocate an environment for invoking the test. */
     {
-        if (!(environ = (char**)calloc(1, sizeof(char**))))
+        if (!(environ = (char**)calloc(1, sizeof(char*))))
```
Comment (jhand2) [inline tests/libc/enc/enc.c]: why change this? They're functionally equivalent and its general practice to use the sizeof the type you're allocating

File: tests/libunwind/enc/enc.c [hunk 10]
```diff
@@ -65,9 +65,9 @@ int test(char test_name[201], uint32_t pid)
     int rval = 1;
     g_pid = pid;
     printf("RUNNING: %s\n", __TEST__);
-    if (!(__environ = (char**)calloc(1, sizeof(char**))))
+    if (!(__environ = (char**)calloc(1, sizeof(char*))))
```
Comment (jhand2) [inline tests/libunwind/enc/enc.c]: same as above

File: tests/memory/enc/basic.c [hunk 11]
```diff
@@ -38,6 +38,8 @@ void test_malloc(void)
     /* Ensure that malloc fails. */
     ptr = (int*)malloc(~((size_t)0));
     OE_TEST(ptr == NULL);
+    if (ptr)
+        free(ptr);
```
Comment (jhand2) [inline tests/memory/enc/basic.c]: This code is unreachable. Do the check/free before OE_TEST

File: tests/memory/enc/basic.c [hunk 12]
```diff
@@ -102,11 +104,8 @@ void test_realloc(void)
 #pragma GCC diagnostic pop
 #endif
     OE_TEST(ptr2 == NULL);
-
-    /* realloc(X, 0) is implementation defined, but we can always free it. */
-    free(ptr);
-    ptr = realloc(NULL, 0);
-    free(ptr);
+    if (ptr2)
+        free(ptr2);
```
Comment (jhand2) [inline tests/memory/enc/basic.c]: Same as above

File: tests/stdc/enc/enc.cpp [hunk 13]
```diff
@@ -274,6 +274,10 @@ int test(char buf1[BUFSIZE], char buf2[BUFSIZE])
     /* Cause malloc() to fail */
     void* p = malloc(1024 * 1024 * 1024);
     OE_TEST(p == NULL);
+
+    if (p)
+        free(p);
```
Comment (jhand2) [inline tests/stdc/enc/enc.cpp]: unreachable

**Comment by jhand2** [comment]
It would appear you accidentally pulled in some extra commits that pull a bunch of stuff into this PR

**Comment by ryanhsu19** [comment]
Hi Jordan,

I will remove the last three ones for now. I changed it because it showed a scan-build error regarding the OE_TEST conditions, but again, OE_TEST runs oe_abort() once the condition is false.

As for the calloc, there is a type-mismatching error. What Alvin and I were thinking was, since char** is an array of char*, we discussed and changed sizeof(char**) to sizeof(char*), and the scan-build error has disappeared since then.

The revocation info part is for the memory leak issues.

I will change them back.

Ryan

File: common/sgx/verifier.c [hunk 14]
```diff
@@ -39,8 +39,10 @@ static oe_result_t _on_unregister(oe_attestation_role_t* context)
 
 static void _free_claim(oe_claim_t* claim)
 {
-    oe_free(claim->name);
-    oe_free(claim->value);
+    if (claim->name != NULL)
```
Comment (dthaler) [inline common/sgx/verifier.c]: The standard free() is specified to be a no-op if NULL is passed, and so should oe_free.
Hence I think this change should not be needed and it should be annotated however free() is annotated so that no warnings are given when NULL is passed.

File: enclave/crypto/cert.c [hunk 15]
```diff
@@ -764,8 +764,8 @@ oe_result_t oe_cert_chain_read_pem(
     result = OE_OK;
 
 done:
-
-    _referent_free(referent);
+    if (referent)
+        _referent_free(referent);
```
Comment (dthaler) [inline enclave/crypto/cert.c]: Same comment here about _referent_free(), it should a no-op if NULL is passed so this change should not be needed.  Fix the header so no warning is generated.

File: enclave/sgx/revocationinfo.c [hunk 16]
```diff
@@ -160,10 +158,12 @@ void oe_free_get_revocation_info_args(oe_get_revocation_info_args_t* args)
         free(args->tcb_issuer_chain);
 
         for (size_t i = 0; i < OE_COUNTOF(args->crl); i++)
-            free(args->crl[i]);
+            if (args->crl[i])
+                free(args->crl[i]);
```
Comment (dthaler) [inline enclave/sgx/revocationinfo.c]: This was already safe per specification of standard free().  Do we know why it generated a warning?  Rather than changing all code that is already safe, it's better to have a fix that stops generating bad warnings.

File: common/sgx/verifier.c (hunk 14 above)
Comment (jhand2) [reply common/sgx/verifier.c]: Indeed both `oe_free` and `free` check for null ptr.

@ryanhsu19 can you share the warning generated for these?

**Comment by ryanhsu19** [comment]
As for enclave/sgx/revocationinfo.c, I made some changes due to previously many double-free warnings (there is only one such warning now and it is for a third-party file), and I included the change this file as I added the in-buffers into the /* free buffers */ section in lines 140-149. Same for common/sgx/verifier.c.

As for the enclave/crypto/cert.c, if you enter from mbedtls_x509_crt_free(referent->crt) and referent is a null pointer, then access to field crt in _referent_free() results in a dereference of a null pointer, therefore the warning. Since this is now the only one I can reconstruct, I will drop verifier.c and keep the in-buffers (to prevent possible memory leak) in revocationinfo.c.

**Comment by jhand2** [comment]
@ryanhsu19 Can you reply to people's comments directly instead of making a new comment to address issues. Its too cumbersome to jump around the PR to find what you're referencing.

File: common/sgx/verifier.c (hunk 14 above)
Comment (ryanhsu19) [reply common/sgx/verifier.c]: I included it because previously scan-build popped up some double-free warnings for many files, but I could not reconstruct it so I will remove it.

File: enclave/crypto/cert.c (hunk 15 above)
Comment (ryanhsu19) [reply enclave/crypto/cert.c]: If you enter from mbedtls_x509_crt_free(referent->crt) and referent is a null pointer, then access to field crt in _referent_free() results in a dereference of a null pointer, therefore the warning.

File: enclave/sgx/revocationinfo.c (hunk 16 above)
Comment (ryanhsu19) [reply enclave/sgx/revocationinfo.c]: I included it because scan-build popped up many double-free warnings. The main purpose for the file edit is to add in-buffers into the free buffer section in lines 140-149 to prevent potential memory leaks.

File: enclave/crypto/cert.c (hunk 15 above)
Comment (jhand2) [reply enclave/crypto/cert.c]: _referent_free does not check for a null pointer. We should fix the implementation to check there.

File: enclave/sgx/revocationinfo.c (hunk 16 above)
Comment (jhand2) [reply enclave/sgx/revocationinfo.c]: So if there are double free warnings, does this really solve them? `free` does not set the pointer to null, so even with this check the same double-free bug would persist.

File: common/sgx/verifier.c (hunk 14 above)
Comment (jhand2) [reply common/sgx/verifier.c]: Fixing double free would require setting `claim->name` and `claim->value` to NULL after free. Not checking for NULL.

Comment (anakrish) [reply common/sgx/verifier.c]: Couple of questions:

1. Why does scanbuild not care that the `claim` variable is not null-checked before dereferencing (`claim-<name`)?

2. Why does scanbuild care that the parameter to oe_free be null checked. It is valid to pass NULL value to functions.

Comment (jhand2) [reply common/sgx/verifier.c]: 1. Scan build only catches bugs that can actually happen in the execution path. So if no code ever passes null to `_free_claim` then scan-build won't warn about it.
2. It doesn't. I don't think this fixes the reported warning.

File: enclave/crypto/cert.c (hunk 15 above)
Comment (ryanhsu19) [reply enclave/crypto/cert.c]: Okay. Working on it now.

Comment (dthaler) [reply enclave/crypto/cert.c]: +1 to fixing _referent_free to check for a null pointer, for consistent with the rest of the *free() functions

File: samples/file-encryptor/enclave/keys.cpp [hunk 17]
```diff
@@ -290,9 +290,17 @@ int ecall_dispatcher::prepare_encryption_header(
...
+        }
```
Comment (jhand2) [inline samples/file-encryptor/enclave/keys.cpp]: Remove

File: tests/libc/enc/enc.c (hunk 9 above)
Comment (jhand2) [reply tests/libc/enc/enc.c]: This is not resolved

File: tests/libunwind/enc/enc.c (hunk 10 above)
Comment (jhand2) [reply tests/libunwind/enc/enc.c]: Also not resolved

File: tests/memory/enc/basic.c [hunk 18]
```diff
@@ -102,11 +102,6 @@ void test_realloc(void)
 #pragma GCC diagnostic pop
 #endif
     OE_TEST(ptr2 == NULL);
-
-    /* realloc(X, 0) is implementation defined, but we can always free it. */
-    free(ptr);
-    ptr = realloc(NULL, 0);
-    free(ptr);
```
Comment (jhand2) [inline tests/memory/enc/basic.c]: Why was this removed?

File: tests/syscall/cpio/commands.c [hunk 19]
```diff
@@ -72,7 +72,7 @@ int oe_lsr(const char* root, oe_strarr_t* paths)
 
     oe_strarr_release(&dirs);
 
-    if (ret != 0)
+    if (ret != 0 && paths != 0)
```
Comment (jhand2) [inline tests/syscall/cpio/commands.c]: should be `paths != NULL`

File: enclave/crypto/cert.c [hunk 20]
```diff
@@ -96,15 +96,18 @@ OE_INLINE void _referent_add_ref(Referent* referent)
 OE_INLINE void _referent_free(Referent* referent)
 {
     /* If this was the last reference, release the object */
-    if (oe_atomic_decrement(&referent->refs) == 0)
+    if (referent)
```
Comment (mingweishih) [inline enclave/crypto/cert.c]: For the code readability, consider do the check and early return instead of introducing nested brackets.
For example, `if (!referent) return;`

File: tests/libc/enc/enc.c (hunk 9 above)
Comment (ryanhsu19) [reply tests/libc/enc/enc.c]: I know char* and char** are both pointers, but the meanings are different, and it does generate 10 related sizeof operand mismatch. Therefore I am wondering why char** is designed to point to an array of one char** element.

File: samples/file-encryptor/enclave/keys.cpp (hunk 17 above)
Comment (ryanhsu19) [reply samples/file-encryptor/enclave/keys.cpp]: Sorry I accidentally added it. I will remove it.

File: tests/syscall/cpio/commands.c (hunk 19 above)
Comment (ryanhsu19) [reply tests/syscall/cpio/commands.c]: Thanks. Will show up in next update

File: tests/memory/enc/basic.c (hunk 18 above)
Comment (ryanhsu19) [reply tests/memory/enc/basic.c]: Reverted it. There is still a problem when attempting to release already-freed ptr at line 106 (assume ptr != NULL and ptr is released at line 100), but I will try to deal with it after other warnings are settled.

File: enclave/crypto/cert.c (hunk 20 above)
Comment (ryanhsu19) [reply enclave/crypto/cert.c]: Got it. I will change it now.

File: common/sgx/verifier.c [hunk 21]
```diff
@@ -118,10 +118,7 @@ static oe_result_t _add_claim(
 
     claim->value = (uint8_t*)oe_malloc(value_size);
     if (claim->value == NULL)
-    {
-        oe_free(claim->name);
```
[the remark this answers is not in the document]
Comment (ryanhsu19) [reply common/sgx/verifier.c]: I removed it because I could not reconstruct the situation and considered it unrelated. There is no scan-build error involved.

File: tests/libc/enc/enc.c (hunk 9 above)
Comment (jhand2) [reply tests/libc/enc/enc.c]: oh, sorry, you're right. Re-resolving :smile:

File: tests/libunwind/enc/enc.c (hunk 10 above)
Comment (jhand2) [reply tests/libunwind/enc/enc.c]: My mistake. Resolving

File: common/sgx/verifier.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (jhand2) [reply common/sgx/verifier.c]: But the oe_free line is in master. So if there is no associated warning you should leave it.

File: samples/file-encryptor/enclave/keys.cpp [hunk 22]
```diff
@@ -290,9 +290,11 @@ int ecall_dispatcher::prepare_encryption_header(
     {
         TRACE_ENCLAVE("password_key");
         for (unsigned int i = 0; i < ENCRYPTION_KEY_SIZE_IN_BYTES; i++)
+        {
             TRACE_ENCLAVE(
                 "password_key[%d] =0x%02x", i, (unsigned int)(password_key[i]));
-        goto exit;
+            goto exit;
```
Comment (jhand2) [inline samples/file-encryptor/enclave/keys.cpp]: Revert this back to what is in master.

Comment (ryanhsu19) [reply samples/file-encryptor/enclave/keys.cpp]: Done

File: tests/memory/enc/basic.c (hunk 18 above)
Comment (ryanhsu19) [reply tests/memory/enc/basic.c]: I thought about it for a while and reverted to the original version.

So there are 3 scan-build warnings, one for double-free and the others for memory leaks. Since the test is intended to ensure that memory allocation fails at the end of each unit test, it is not likely to change OE_TEST(ptr == NULL) and OE_TEST(ptr2 == NULL).

The double-free warning occurs because scan-build does not recognize oe_abort(), or it thinks that abort() is different than exit(). Because if ptr is really released at line 100, it means that ptr2 is successfully reallocated, therefore failing the test and leaving the rest of the code unreachable.

File: common/sgx/verifier.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (ryanhsu19) [reply common/sgx/verifier.c]: It does show "using freed memory" error in _free_claim() after I add it back, but I remember you said that oe_free(param) checks if param is null before it does anything right?

[the remark this answers is not in the document]
Comment (jhand2) [reply common/sgx/verifier.c]: It does, but claim->name is never set to null after the free. So before returning just do `claim->name = NULL;`

[the remark this answers is not in the document]
Comment (ryanhsu19) [reply common/sgx/verifier.c]: Oh, that's right. I just added it back.

**Comment by jhand2** [comment]
Thanks! Please remove merge commits and squash commit history.
